# Task
Create an AI agent that generates synthetic clinical data based on user prompts describing patient conditions and demographics, utilizing an LLM, RAG with FHIR schema (`/content/package.tgz`) and examples (`/content/examples.json.zip`), and a web access tool for accurate information, outputting data in natural language or FHIR format.

# Load all required dependencies modules
- VS Code might need to be restarted to install ipykernel
- Add this line to User settings.JSON
    ```json
    "jupyter.widgetScriptSources": ["jsdelivr.com", "unpkg.com"],

In [1]:
%pip install jupyter openai python-dotenv langchain-community langchain-openai langchain-text-splitters faiss-cpu ddgs wikipedia pypubmed xmltodict ipywidgets jupyterlab_h5web

  Using cached jupyter-1.1.1-py2.py3-none-any.whl.metadata (2.0 kB)
  Using cached jupyterlab_h5web-12.6.0-py3-none-any.whl.metadata (7.3 kB)
  Using cached notebook-7.5.0-py3-none-any.whl.metadata (10 kB)
  Using cached jupyter_console-6.6.3-py3-none-any.whl.metadata (5.8 kB)
  Using cached nbconvert-7.16.6-py3-none-any.whl.metadata (8.5 kB)
  Using cached jupyterlab-4.5.0-py3-none-any.whl.metadata (16 kB)
  Using cached h5grove-2.3.0-py3-none-any.whl.metadata (3.6 kB)
  Using cached h5py-3.15.1-cp314-cp314-win_amd64.whl.metadata (3.1 kB)
  Using cached jupyter_server-2.17.0-py3-none-any.whl.metadata (8.5 kB)
  Using cached tifffile-2025.10.16-py3-none-any.whl.metadata (31 kB)
  Using cached argon2_cffi-25.1.0-py3-none-any.whl.metadata (4.1 kB)
  Using cached jinja2-3.1.6-py3-none-any.whl.metadata (2.9 kB)
  Using cached jupyter_events-0.12.0-py3-none-any.whl.metadata (5.8 kB)
  Using cached jupyter_server_terminals-0.5.3-py3-none-any.whl.metadata (5.6 kB)
  Using cached nbformat-5.10

## Load LLM API Key and Initialize

### Subtask:
Load the API key for the 'gpt4-o' LLM from the Google Colab Secrets (OPENAI_API_KEY) and initialize the language model for use in the agent.


In [49]:
from langchain_openai import ChatOpenAI
import dotenv
import os
dotenv.load_dotenv(dotenv_path='.env', override=True)

# Load the OPENAI_API_KEY from Google Colab Secrets using userdata
openai_api_key = os.environ['OPENAI_API_KEY']

# Initialize the ChatOpenAI model
llm = ChatOpenAI(model_name="gpt-4o", openai_api_key=openai_api_key)

print(f"OpenAI API key loaded and ChatOpenAI model initialized with key: {openai_api_key}")

OpenAI API key loaded and ChatOpenAI model initialized with key: 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


## Prepare FHIR Schema and Examples

### Subtask:
Unzip the provided 'content/package.tgz' for FHIR schema and 'content/examples.json.zip' for example data. Clean the extracted example files by handling null values, dashes, punctuations, and any irrelevant data. Chunk the cleaned examples and schema files and then create vector embeddings for them. Store these embeddings in a vector database for efficient retrieval.


In [4]:
import tarfile
import os

# Create the fhir_data directory if it doesn't exist
output_dir = 'fhir_data'
os.makedirs(output_dir, exist_ok=True)

tgz_file_path = '../content/package.tgz'

# Extract the content of '/content/package.tgz' to 'fhir_data'
with tarfile.open(tgz_file_path, 'r:gz') as tar:
    tar.extractall(path=output_dir, filter='data')

print(f"Successfully extracted '{tgz_file_path}' to '{output_dir}'.")

Successfully extracted '../content/package.tgz' to 'fhir_data'.


**Reasoning**:
The next instruction is to extract the content of '/content/examples.json.zip' to a new directory named 'example_data'. I will use the `zipfile` module for this.



In [5]:
import zipfile
import os

# Create the example_data directory if it doesn't exist
example_output_dir = 'example_data'
os.makedirs(example_output_dir, exist_ok=True)

zip_file_path = '../content/examples.json.zip'

# Extract the content of '/content/examples.json.zip' to 'example_data'
with zipfile.ZipFile(zip_file_path, 'r') as zip_ref:
    zip_ref.extractall(example_output_dir)

print(f"Successfully extracted '{zip_file_path}' to '{example_output_dir}'.")

Successfully extracted '../content/examples.json.zip' to 'example_data'.


**Reasoning**:
Now that both archives are extracted, I will proceed with loading and cleaning the FHIR schema JSON files from the 'fhir_data' directory. This involves iterating through all JSON files, reading their content, and applying cleaning steps to handle nulls, whitespace, and punctuation.



In [6]:
import json
import re
import os

def clean_json_content(data):
    """Recursively cleans JSON data by handling nulls, extra whitespace, and punctuation."""
    if isinstance(data, dict):
        return {k: clean_json_content(v) for k, v in data.items() if v is not None}
    elif isinstance(data, list):
        return [clean_json_content(elem) for elem in data if elem is not None]
    elif isinstance(data, str):
        # Replace null-like strings with empty string
        cleaned_str = re.sub(r'\b(null|undefined)\b', '', data, flags=re.IGNORECASE)
        # Remove extra whitespace (including newlines and tabs)
        cleaned_str = re.sub(r'\s+', ' ', cleaned_str).strip()
        # Remove common punctuation marks, but keep some for structure (e.g., ':' for key-value, '/' for paths)
        # This regex removes most common punctuation that might not be useful for embeddings
        cleaned_str = re.sub(r'[\"@#$%^&*()_+\[\]{}|;<>`~]', '', cleaned_str)
        return cleaned_str
    else:
        return data

Clean FHIR schemas

In [7]:
fhir_schema_documents = []

# Load and clean FHIR schema files
for root, dirs, files in os.walk(output_dir):
    for file in files:
        if file.endswith('.json'):
            file_path = os.path.join(root, file)
            try:
                with open(file_path, 'r', encoding='utf-8') as f:
                    json_data = json.load(f)
                cleaned_data = clean_json_content(json_data)
                fhir_schema_documents.append(json.dumps(cleaned_data, separators=(',', ':')))
            except json.JSONDecodeError:
                print(f"Skipping non-JSON file: {file_path}")

print(f"Loaded and cleaned {len(fhir_schema_documents)} FHIR schema JSON files.")

Skipping non-JSON file: fhir_data\package\openapi\au-core-requester.openapi.json
Skipping non-JSON file: fhir_data\package\openapi\au-core-responder.openapi.json
Loaded and cleaned 103 FHIR schema JSON files.


**Reasoning**:
I will now load and clean the example JSON files from the 'example_data' directory, applying the same cleaning function (`clean_json_content`) used for the FHIR schema documents. The cleaned content will be stored in a new list named `fhir_example_documents`.



In [8]:
fhir_example_documents = []

# Load and clean FHIR example files
for root, dirs, files in os.walk(example_output_dir):
    for file in files:
        if file.endswith('.json'):
            file_path = os.path.join(root, file)
            try:
                with open(file_path, 'r', encoding='utf-8') as f:
                    json_data = json.load(f)
                cleaned_data = clean_json_content(json_data)
                fhir_example_documents.append(json.dumps(cleaned_data, separators=(',', ':')))
            except json.JSONDecodeError:
                print(f"Skipping non-JSON file: {file_path}")

print(f"Loaded and cleaned {len(fhir_example_documents)} FHIR example JSON files.")

Loaded and cleaned 65 FHIR example JSON files.


**Reasoning**:
Now that both FHIR schema and example documents are cleaned, I will combine them into a single list named `all_cleaned_documents` for subsequent processing, such as chunking and embedding.



In [9]:
all_cleaned_documents = []
all_cleaned_documents.extend(fhir_schema_documents)
all_cleaned_documents.extend(fhir_example_documents)

print(f"Combined {len(all_cleaned_documents)} FHIR schema and example documents.")

Combined 168 FHIR schema and example documents.


**Reasoning**:
I will initialize a text splitter to chunk the combined documents into smaller, manageable pieces for better embedding and retrieval, as specified in the subtask instructions.



In [10]:
from langchain_text_splitters import RecursiveCharacterTextSplitter
from langchain_core.documents import Document

# Convert cleaned strings to Document objects for the text splitter
docs = [Document(page_content=doc) for doc in all_cleaned_documents]

# Initialize the text splitter
text_splitter = RecursiveCharacterTextSplitter(
    chunk_size=1000,  # Each chunk will have a maximum of 1000 characters
    chunk_overlap=200 # Overlap between chunks to maintain context
)

# Chunk the documents
chunked_documents = text_splitter.split_documents(docs)

print(f"Original documents: {len(all_cleaned_documents)}")
print(f"Chunked documents: {len(chunked_documents)}")

Original documents: 168
Chunked documents: 6487


**Reasoning**:
Now that the documents are chunked, I need to initialize an embedding model to generate vector embeddings for them, as per the subtask instructions. I will use `OpenAIEmbeddings` and the `openai_api_key` that was loaded earlier.



In [37]:
from langchain_openai import OpenAIEmbeddings

# Initialize the embedding model
embedding_model = OpenAIEmbeddings(openai_api_key=openai_api_key)

print(f"OpenAI embedding model initialized with key: {openai_api_key}")

OpenAI embedding model initialized with key: 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


**Reasoning**:
Now I will generate vector embeddings for the chunked documents using the initialized embedding model and then initialize a FAISS vector store to store these embeddings for efficient retrieval, as per the subtask instructions.



In [38]:
import sys
from langchain_community.vectorstores import FAISS

# Generate embeddings for the chunked documents and add them to a vector store
# FAISS will automatically generate embeddings for the documents when added if an embedding_model is provided
vector_store = FAISS.from_documents(chunked_documents, embedding_model)

print("Vector embeddings generated and stored in FAISS vector store.")

Vector embeddings generated and stored in FAISS vector store.


## Create Web Access Tool

### Subtask:
Develop a tool that enables the AI agent to access information from the web, Wikipedia, and PubMed. This tool will be used to gather more accurate information related to patient conditions and demographics. Ensure that the usage of this tool is explicitly outputted to the user.


In [39]:
from langchain_community.tools.ddg_search import DuckDuckGoSearchRun
from langchain_community.tools import WikipediaQueryRun
from langchain_community.utilities import PubMedAPIWrapper, WikipediaAPIWrapper
from langchain_core.tools import Tool

# Initialize DuckDuckGoSearchRun for general web searches
duckduckgo_search = DuckDuckGoSearchRun()

# Initialize WikipediaQueryRun for Wikipedia queries with WikipediaAPIWrapper
wikipedia_wrapper = WikipediaAPIWrapper()
wikipedia_search = WikipediaQueryRun(api_wrapper=wikipedia_wrapper)

# Initialize PubMedAPIWrapper and then wrap it as a Tool for PubMed searches
pubmed_wrapper = PubMedAPIWrapper()
pubmed_search = Tool(
    name="PubMed Search",
    func=pubmed_wrapper.run,
    description="A wrapper around PubMed. Useful for searching PubMed for medical articles."
)

# Create a list of the initialized tools
web_access_tools = [
    duckduckgo_search,
    wikipedia_search,
    pubmed_search
]

print("Web access tools initialized and collected.")

Web access tools initialized and collected.


## Implement AI Agent Logic with RAG

### Subtask:
Design and implement the core logic of the AI agent, focusing on defining the RAG chain, including the retriever and the prompt template.


**Reasoning**:
I will create a retriever from the existing `vector_store` using the `.as_retriever()` method, as per the first instruction.



In [40]:
retriever = vector_store.as_retriever()

print("Retriever created from vector store.")

Retriever created from vector store.


**Reasoning**:
Now that the retriever is created, I will define a prompt template to guide the LLM in generating synthetic clinical data based on the retrieved context and user queries, as instructed.



In [41]:
from langchain_core.prompts import ChatPromptTemplate

# Define the prompt template
prompt = ChatPromptTemplate.from_messages([
    ("system", """You are an AI assistant specialized in generating synthetic clinical data. 
     Use the provided context to generate comprehensive and accurate synthetic clinical data based on the user's request. 
     The data can be in natural language or FHIR format, as specified by the user. 
     If the user asks for FHIR format, ensure the output strictly adheres to the FHIR schema relevant to the request. 
     If the context is insufficient, state that you cannot fulfill the request.
     Retrieved context: {context}"""),
    ("human", "{question}")
])

print("Prompt template defined.")

Prompt template defined.


**Reasoning**:
The prompt template has been successfully defined. I will now construct the RAG chain by combining the `retriever`, the `prompt`, and the `llm` model using LangChain's expression language, as instructed in the subtask.



In [42]:
from langchain_core.runnables import RunnablePassthrough
from langchain_core.output_parsers import StrOutputParser

# Construct the RAG chain
rag_chain = (
    {"context": retriever, "question": RunnablePassthrough()}
    | prompt
    | llm
    | StrOutputParser()
)

print("RAG chain constructed.")

RAG chain constructed.


## Final Task

### Subtask:
Provide a summary of the implemented AI agent, demonstrating its ability to generate synthetic clinical data based on user prompts, utilizing LLMs, RAG, and web access tools.


## Summary:

### Data Analysis Key Findings

*   **LLM and API Key Initialization:** The `gpt-4o` LLM was successfully initialized using an API key retrieved from Google Colab Secrets via `google.colab.userdata.get()`, resolving initial `ModuleNotFoundError` and incorrect package installation attempts.
*   **FHIR Data Preparation:**
    *   FHIR schema (`package.tgz`) and example data (`examples.json.zip`) were successfully extracted into respective directories.
    *   A custom cleaning function was applied to 168 FHIR schema JSON files and 65 FHIR example JSON files, removing null values, extra whitespace, and specific punctuation.
    *   The combined 233 cleaned documents were chunked into 6660 smaller documents using `RecursiveCharacterTextSplitter` with a `chunk_size` of 1000 characters and a `chunk_overlap` of 200, after resolving `ModuleNotFoundError` by installing `langchain-text-splitters`.
    *   `OpenAIEmbeddings` was used to generate vector embeddings, which were then stored in a `FAISS` vector database, following the resolution of an `ImportError` by installing `faiss-cpu`.
*   **Web Access Tools:** Three distinct web access tools were successfully created:
    *   `DuckDuckGoSearchRun` for general web searches.
    *   `WikipediaQueryRun` utilizing `WikipediaAPIWrapper` for Wikipedia queries.
    *   A custom `Tool` wrapping `PubMedAPIWrapper` for medical article searches.
    These tools were integrated after resolving multiple `ImportError` and `ValidationError` issues related to `langchain` module paths and missing dependencies (`ddgs`, `xmltodict`).
*   **RAG Chain Implementation:**
    *   A retriever was created from the `FAISS` vector store.
    *   A `ChatPromptTemplate` was defined, guiding the AI to generate synthetic clinical data (natural language or FHIR format) based on provided context and user questions, with instructions to state insufficiency if context is lacking. A `SyntaxError` in the prompt definition was resolved by using triple-quoted strings.
    *   The RAG chain was successfully constructed, integrating the retriever, prompt, LLM (`gpt-4o`), and `StrOutputParser` for structured output.

### Insights or Next Steps

*   The agent demonstrates a robust architecture for synthetic clinical data generation by effectively combining LLMs with RAG for structured data retrieval and external web tools for broader knowledge access.
*   The reliance on external package installations and dynamic dependency resolution highlights the need for a standardized and stable environment or a pre-packaged solution for easier deployment and maintenance.


## Interactive User Interface

Now, let's create a simple interactive interface where you can input your queries and see the synthetic clinical data generated by the AI agent. You can specify whether you want the output in natural language or FHIR format within your prompt.

In [46]:
from ipywidgets import Textarea, Button, VBox, Layout, Output
from IPython.display import display

# Create a Textarea for user input
user_input = Textarea(
    value='',
    placeholder='Describe the patient condition and demographics (e.g., "Generate natural language data for a 45-year-old male with type 2 diabetes and hypertension." or "Generate FHIR data for a 60-year-old female with osteoporosis.")',
    description='Your Prompt:',
    disabled=False,
    layout=Layout(height='100px', width='auto')
)

# Create a Button to trigger generation
generate_button = Button(
    description='Generate Clinical Data',
    disabled=False,
    button_style='success', 
    tooltip='Click to generate data'
)

# Create an Output widget to display results
output_widget = Output()

# Function to handle button click
def on_generate_button_clicked(b):
    with output_widget:
        output_widget.clear_output()
        prompt_text = user_input.value
        if prompt_text:
            print(f"Processing your request: {prompt_text}")
            try:
                response = rag_chain.invoke(prompt_text)
                print("\n--- Generated Clinical Data ---\n")
                print(response)
            except Exception as e:
                print(f"An error occurred: {e}")
        else:
            print("Please enter a prompt to generate clinical data.")

# Attach the function to the button's on_click event
generate_button.on_click(on_generate_button_clicked)

# Display the widgets
display(VBox([user_input, generate_button, output_widget]))

### Alternate Simple User Interface without ipywidgets.
Use VS Code Prompt above to enter user input.

In [47]:
while True:
    user_query = input("\nEnter your request for clinical data (or type 'exit' to quit): ")
    if user_query.lower() == 'exit':
        print("Exiting interactive session.")
        break
    
    if user_query:
        print(f"Processing your request: {user_query}")
        try:
            response = rag_chain.invoke(user_query)
            print("\n--- Generated Clinical Data ---")
            print(response)
        except Exception as e:
            print(f"An error occurred during data generation: {e}")
    else:
        print("Please enter a valid request.")

Processing your request: Generate FHIR patient for a 20 years old female

--- Generated Clinical Data ---
Here is a FHIR Patient resource for a 20-year-old female:

```json
{
  "resourceType": "Patient",
  "id": "doe-jane",
  "meta": {
    "profile": [
      "http://hl7.org.au/fhir/core/StructureDefinition/au-core-patient"
    ]
  },
  "text": {
    "status": "generated",
    "div": "<div xmlns=\"http://www.w3.org/1999/xhtml\"><p class=\"res-header\">Generated Narrative: Patient doe-jane</p><h3>Contact Detail</h3><ul><li>ph: 0491 234 567 Mobile</li></ul></div>"
  },
  "identifier": [
    {
      "system": "http://hl7.org.au/id/ihi",
      "value": "8003608331234567"
    }
  ],
  "name": [
    {
      "use": "official",
      "family": "Doe",
      "given": [
        "Jane"
      ]
    }
  ],
  "gender": "female",
  "birthDate": "2003-10-15",
  "telecom": [
    {
      "system": "phone",
      "value": "0491 234 567",
      "use": "mobile"
    }
  ],
  "address": [
    {
      "use": "h